### PR TITLE
BZ 64752 - Add GraphQL/HTTP Request Sampler 

### PR DESCRIPTION
**Pull request by woonsan**
## Description

New GraphQL HTTP Request sampler, just a GUI variation of the HTTP Request sampler, was added.

Changes:
- Added `GraphQLHTTPSamplerGui` (extending `HTTPSamplerGui`) and `GraphQLUrlConfigGui` (extending `UrlConfigGui`) in order to customize some UI elements (e.g, hiding irrelevant UI elements in GraphQL requests, adding more convenient GraphQL request content UI elements, etc.), with minimal and backward-compatible refactoring in the base classes to make the customizations possible.
- Added com.google.code.gson:gson library for JSON parsing and serialization.
- Site documentation updates.

## Motivation and Context

I have used HTTP Request sampler for GraphQL testing with setting request body with escaped json strings like this:

```
{"operationName":null,"variables":{},"query":"{\n fineSomethings(text:
\"\", offset: 0, limit: 200) {\n offset\n limit\n count\n total\n
items {\n ... }\n }\n }\n }\n}\n"}
```

It is very hard to read and edit the GraphQL parameter content, especially comparing with GraphQL client tools such as graphiql playground where you can view or edit `Query` and `Variables` directly without any escaping.

It would be nice if we have a GraphQL HTTP Request sampler configuration for better UX.

ASF Bugzilla ticket: https://bz.apache.org/bugzilla/show_bug.cgi?id=64752

## How Has This Been Tested?

In order to test it, we need a GraphQL server. We can test the feature with this open source demo GraphQL server:
- https://github.com/apollographql/starwars-server

#### Basic Tests

- Clone the https://github.com/apollographql/starwars-server and install/start it like the following:
```
git clone https://github.com/apollographql/starwars-server
cd starwars-server
npm install
npm start
```
The demo server will be available at http://localhost:8080/graphql.
(If you want, you can also visit http://localhost:8080/graphql in your browser to see the `graphiql` tool.)
- Run JMeter and open [xdocs/demos/SimpleGraphQLTestPlan.jmx](xdocs/demos/SimpleGraphQLTestPlan.jmx) and start the test.
- Verify if the response is okay in the 'View Results Tree' listener.

#### Recording Tests

With "Recording Template with Think Time" for example, you may record GraphQL HTTP requests by making queries against the same demo GraphQL server mentioned above. (Tip: I had to use another hostname, e.g. http://laptop1:8080/graphql, instead of 'localhost:8080' or '127.0.0.1:8080' which don't use the proxy setting at all in my FireFox.)

- By default, when you record http requests following the instructions, any GraphQL over HTTP requests will be detected automatically and captured with the `GraphQLHTTPSamplerGui` UI by default.
- However, the automatic GraphQL HTTP Request detection cannot be perfect because GraphQL over HTTP just uses "application/json" as content type and there's nothing really differentiating from other http requests clearly, by specification.
- Therefore, the PR includes the best efforts in the automatic detection by checking body/parameter types and content.
- In case where GraphQL-like requests are recorded for some reason and the auto-detection is undesirable, users can turn off the auto-detection mode by unchecking the checkbox ("[x] Detect GraphQL Request") in the proxy control GUI.
- If turned off (turned on by default), no extra modification for GraphQL, but just recorded as normal http requests.

### Potential affecting area

HTTP Request and AJP Request samplers are affected.

## Screenshots (if appropriate):

- [xdocs/images/screenshots/graphql-http-request.png](xdocs/images/screenshots/graphql-http-request.png)
GraphQL HTTP Request
- [xdocs/images/screenshots/graphql-http-request-vars.png](xdocs/images/screenshots/graphql-http-request-vars.png)
GraphQL Variables

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

File: src/core/src/main/resources/org/apache/jmeter/resources/messages_de.properties [hunk 1]
```diff
@@ -224,6 +224,11 @@ graph_results_ms=Millisekunden (ms)
 graph_results_no_samples=Anzahl der Proben
 graph_results_throughput=Durchsatz
 graph_results_title=Ergebnisse
+graphql_http_sampler_title=GraphQL HTTP Request
```
Comment (pmouawad) [inline src/core/src/main/resources/org/apache/jmeter/resources/messages_de.properties:227]: You don't need to provide those messages_XX.properties if you don't translate them.
Only messages.properties and messages_fr.properties are mandatory

File: lib/aareadme.txt [hunk 2]
```diff
@@ -106,6 +106,11 @@ groovy-3.0.3
 ----------------------
 Advised scripting language for JSR223 Test Elements
 
+gson-2.8.6
+----------------------
+https://github.com/google/gson
```
Comment (pmouawad) [inline lib/aareadme.txt:111]: Wasn't it possible to use Jackson which is already a dependency of JMeter ?
https://www.baeldung.com/jackson-vs-gson

**Comment by pmouawad** [comment]
Hello @woonsan ,
Thanks for this great contribution.
I have added few notes.

Thanks
Regards

File: lib/aareadme.txt (hunk 2 above)
Comment (woonsan) [reply lib/aareadme.txt]: Fixed: f9a10107a
Thanks!

File: src/core/src/main/resources/org/apache/jmeter/resources/messages_de.properties (hunk 1 above)
Comment (woonsan) [reply src/core/src/main/resources/org/apache/jmeter/resources/messages_de.properties]: Fixed: fcf3ed2fc (Just removed the unnecessary, untranslated ones).
Thanks!

File: src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties [hunk 3]
```diff
@@ -443,6 +443,11 @@ graph_results_ms=ms
 graph_results_no_samples=Nombre d'échantillons
 graph_results_throughput=Débit
 graph_results_title=Graphique de résultats
+graphql_http_sampler_title=GraphQL HTTP Request
```
Comment (pmouawad) [inline src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties:446]: GraphQL HTTP Request
graphql_request_info=Requête HTTP GraphQL
graphql_operation_name=Nom de l'opération
graphql_query=Requête
graphql_variables=Variables

**Comment by pmouawad** [comment]
Hello @woonsan ,

Did you test recording of GraphQL requests ?
Which UI is used ? I guess it's HTTP Request no ?

This might be adjustable in DefaultSamplerCreator#createSampler

Thanks
Regards

File: src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties (hunk 3 above)
Comment (woonsan) [reply src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties]: Thanks @pmouawad ! :-)
I've pushed a fix: a4ef1e3d7f47989e55ac9109c8adfcaebe535134

**Comment by woonsan** [comment]
Hello @pmouawad ,
Thanks again for the reviews and suggestions!
I've pushed fixes to support test recording, too. PR description was updated accordingly. Please take a look: 790e46c03, e25437920.
And would you mind giving me French translation for these, too?

```
detect_graphql_request=Detect GraphQL Request
proxy_sampler_graphql_settings=GraphQL HTTP Sampler settings
```

Kind regards,

Woonsan

File: src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties [hunk 4]
```diff
@@ -288,6 +288,7 @@ deltest=Suppression
 deref=Déréférencement des alias
 description=Description
 detail=Détail
+detect_graphql_request=Detect GraphQL Request
```
Comment (ubikloadpack) [inline src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties:291]: Détecter les requêtes GraphQL

File: src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties [hunk 5]
```diff
@@ -853,6 +859,7 @@ proxy_headers=Capturer les entêtes HTTP
 proxy_pause_http_sampler=Créer une nouvelle transaction après la requête (ms) \:
 proxy_recorder_dialog=Enregistreur\: Contrôle des transactions
 proxy_regex=Correspondance des variables par regex ?
+proxy_sampler_graphql_settings=GraphQL HTTP Sampler settings
```
Comment (ubikloadpack) [inline src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties:862]: Configuration de la requête GraphQL

Comment (woonsan) [reply src/core/src/main/resources/org/apache/jmeter/resources/messages_fr.properties]: Thanks, @ubikloadpack ! Applied.

**Pull request merged**